File: typescript/src/components.ts
import type { ComponentIdentifier, HttpMessage, SFVParams } from "./types.js";
import { MissingComponentError } from "./errors.js";
import {
  newSFVParams,
  sfvParamsSet,
  sfvParamsGet,
  parseDictionary,
  serializeBareItem,
  serializeInnerList,
  serializeParams,
  uint8ToBase64,
} from "./sfv.js";

/** Create a simple component identifier. */
export function component(name: string): ComponentIdentifier {
  return { name: name.toLowerCase() };
}

/** Create a component identifier with custom parameters. */
export function componentWithParams(
  name: string,
  params: SFVParams,
): ComponentIdentifier {
  return { name: name.toLowerCase(), params };
}

/** Create a @query-param component identifier. */
export function queryParam(name: string): ComponentIdentifier {
  return { name: "@query-param", params: sfvParamsSet(newSFVParams(), "name", name) };
}

/** Create a component identifier with the ;key parameter for extracting a single member from a Dictionary Structured Field header. */
export function componentWithKey(name: string, key: string): ComponentIdentifier {
  return { name: name.toLowerCase(), params: sfvParamsSet(newSFVParams(), "key", key) };
}

/** Create a component identifier with the ;req flag for response signatures. */
export function componentReq(name: string): ComponentIdentifier {
  return { name: name.toLowerCase(), params: sfvParamsSet(newSFVParams(), "req", true) };
}

/** Create a component identifier with both ;req and ;key parameters for response signatures that bind to a specific dictionary member from the request. */
export function componentReqWithKey(name: string, key: string): ComponentIdentifier {
  return { name: name.toLowerCase(), params: sfvParamsSet(sfvParamsSet(newSFVParams(), "req", true), "key", key) };
}

function hasParam(c: ComponentIdentifier, key: string): boolean {
  return c.params !== undefined && c.params.keys.includes(key);
}

function paramString(c: ComponentIdentifier, key: string): string {
  if (!c.params) return "";
  const v = sfvParamsGet(c.params, key);
  return typeof v === "string" ? v : "";
}

/** Extract a component value from a message. */
export function extractComponent(
  c: ComponentIdentifier,
  msg: HttpMessage,
  reqMsg?: HttpMessage,
): string {
  // ;req: extract from the associated request
  if (hasParam(c, "req")) {
    if (!reqMsg) {
      throw new MissingComponentError(
        `;req specified but no request message available`,
      );
    }
    const stripped: ComponentIdentifier = { name: c.name };
    if (c.params) {
      const sp = c.params.keys
        .filter((k) => k !== "req")
        .reduce(
          (p, k) => sfvParamsSet(p, k, c.params!.values[k]),
          newSFVParams(),
        );
      if (sp.keys.length > 0) stripped.params = sp;
    }
    return extractComponent(stripped, reqMsg);
  }

  if (c.name.startsWith("@")) {
    return extractDerived(c, msg);
  }
  return extractHeader(c, msg);
}

function extractDerived(c: ComponentIdentifier, msg: HttpMessage): string {
  switch (c.name) {
    case "@method":
      if (!msg.isRequest) throw new MissingComponentError("@method on response");
      return (msg.method ?? "").toUpperCase();

    case "@target-uri":
      if (!msg.isRequest) throw new MissingComponentError("@target-uri on response");
      return msg.url!.href;

    case "@authority":
      if (!msg.isRequest) throw new MissingComponentError("@authority on response");
      return extractAuthority(msg.url!);

    case "@scheme":
      if (!msg.isRequest) throw new MissingComponentError("@scheme on response");
      return msg.url!.protocol.replace(":", "").toLowerCase();

    case "@request-target":
      if (!msg.isRequest) throw new MissingComponentError("@request-target on response");
      return extractRequestTarget(msg.url!);

    case "@path":
      if (!msg.isRequest) throw new MissingComponentError("@path on response");
      return msg.url!.pathname || "/";

    case "@query":
      if (!msg.isRequest) throw new MissingComponentError("@query on response");
      return extractQuery(msg.url!);

    case "@query-param": {
      if (!msg.isRequest) throw new MissingComponentError("@query-param on response");
      const name = paramString(c, "name");
      if (!name) throw new MissingComponentError("@query-param requires ;name parameter");
      return extractQueryParam(msg.url!, name);
    }

    case "@status":
      if (msg.isRequest) throw new MissingComponentError("@status on request");
      return String(msg.statusCode ?? 0);

    default:
      throw new MissingComponentError(`unknown derived component ${c.name}`);
  }
}

function extractHeader(c: ComponentIdentifier, msg: HttpMessage): string {
  if (hasParam(c, "sf")) {
    return extractHeaderSF(c, msg);
  }
  if (hasParam(c, "bs")) {
    return extractHeaderBS(c, msg);
  }
  if (hasParam(c, "key")) {
    return extractHeaderKey(c, msg);
  }

  const values = msg.headerValues(c.name);
  if (values.length === 0) {
    throw new MissingComponentError(`header "${c.name}" not present`);
  }
  return values.map((v) => v.trim()).join(", ");
}

function extractHeaderSF(c: ComponentIdentifier, msg: HttpMessage): string {
  const values = msg.headerValues(c.name);
  if (values.length === 0) {
    throw new MissingComponentError(`header "${c.name}" not present`);
  }
  return values.map((v) => v.trim()).join(", ");
}

function extractHeaderBS(c: ComponentIdentifier, msg: HttpMessage): string {
  const values = msg.headerValues(c.name);
  if (values.length === 0) {
    throw new MissingComponentError(`header "${c.name}" not present`);
  }
  return values
    .map((v) => ":" + uint8ToBase64(new TextEncoder().encode(v.trim())) + ":")
    .join(", ");
}

function extractHeaderKey(c: ComponentIdentifier, msg: HttpMessage): string {
  const key = paramString(c, "key");
  if (!key) throw new MissingComponentError(";key parameter is empty");

  const values = msg.headerValues(c.name);
  if (values.length === 0) {
    throw new MissingComponentError(`header "${c.name}" not present`);
  }

  const combined = values.join(", ");
  const members = parseDictionary(combined);
  for (const m of members) {
    if (m.key === key) {
      if (m.innerList) return serializeInnerList(m.innerList);
      if (m.item) {
        let result = serializeBareItem(m.item.value);
        if (m.item.params && m.item.params.keys.length > 0) {
          result += serializeParams(m.item.params);
        }
        return result;
      }
    }
  }
  throw new MissingComponentError(
    `key "${key}" not found in dictionary header "${c.name}"`,
  );
}

function extractAuthority(url: URL): string {
  const host = url.hostname.toLowerCase();
  const port = url.port;
  if (!port) return host;
  // Omit default ports
  if (
    (url.protocol === "http:" && port === "80") ||
    (url.protocol === "https:" && port === "443")
  ) {
    return host;
  }
  return host + ":" + port;
}

function extractRequestTarget(url: URL): string {
  const path = url.pathname || "/";
  const search = url.search;
  return path + search;
}

function extractQuery(url: URL): string {
  // Always include the leading ?
  if (!url.search) return "?";
  return url.search;
}

function extractQueryParam(url: URL, name: string): string {
  const params = url.searchParams;
  if (!params.has(name)) {
    throw new MissingComponentError(`query parameter "${name}" not present`);
  }
  return params.get(name)!;
}

File: golang/components.go
package httpsig

import (
	"encoding/base64"
	"fmt"
	"net/url"
	"strings"
)

// ComponentIdentifier identifies a component to include in a signature base.
type ComponentIdentifier struct {
	Name   string     // lowercase header name or derived component (e.g. "@method", "content-type")
	Params *SFVParams // optional parameters (sf, key, bs, req, name)
}

// Component creates a ComponentIdentifier with no parameters.
func Component(name string) ComponentIdentifier {
	return ComponentIdentifier{Name: strings.ToLower(name)}
}

// ComponentWithParams creates a ComponentIdentifier with the given parameters.
func ComponentWithParams(name string, params *SFVParams) ComponentIdentifier {
	return ComponentIdentifier{Name: strings.ToLower(name), Params: params}
}

// QueryParam creates a @query-param component identifier for the named parameter.
func QueryParam(name string) ComponentIdentifier {
	p := NewSFVParams()
	p.Set("name", name)
	return ComponentIdentifier{Name: "@query-param", Params: p}
}

// ComponentWithKey creates a component identifier with the ;key parameter for
// extracting a single member from a Dictionary Structured Field header.
// For example, ComponentWithKey("signature", "sig1") extracts the "sig1" entry
// from the Signature dictionary header.
func ComponentWithKey(name string, key string) ComponentIdentifier {
	p := NewSFVParams()
	p.Set("key", key)
	return ComponentIdentifier{Name: strings.ToLower(name), Params: p}
}

// ComponentReq creates a component identifier with the ;req flag for request-bound
// response signatures.
func ComponentReq(name string) ComponentIdentifier {
	p := NewSFVParams()
	p.Set("req", true)
	return ComponentIdentifier{Name: strings.ToLower(name), Params: p}
}

// isDerived returns true if this is a derived component (starts with @).
func (c ComponentIdentifier) isDerived() bool {
	return len(c.Name) > 0 && c.Name[0] == '@'
}

// hasParam checks if a parameter is set.
func (c ComponentIdentifier) hasParam(key string) bool {
	if c.Params == nil {
		return false
	}
	_, ok := c.Params.Get(key)
	return ok
}

// paramString returns a string parameter value or empty string.
func (c ComponentIdentifier) paramString(key string) string {
	if c.Params == nil {
		return ""
	}
	v, ok := c.Params.Get(key)
	if !ok {
		return ""
	}
	s, _ := v.(string)
	return s
}

// extractComponent extracts the value of a component from a message.
// For request-bound components (;req), reqMsg is the associated request.
func extractComponent(c ComponentIdentifier, msg Message, reqMsg Message) (string, error) {
	// If ;req is set, extract from the request message
	if c.hasParam("req") {
		if reqMsg == nil {
			return "", fmt.Errorf("%w: ;req specified but no request message available", ErrMissingComponent)
		}
		// Build a component without the ;req param to extract from the request
		stripped := ComponentIdentifier{Name: c.Name}
		if c.Params != nil {
			sp := NewSFVParams()
			for _, k := range c.Params.Keys {
				if k != "req" {
					sp.Set(k, c.Params.Values[k])
				}
			}
			if sp.Len() > 0 {
				stripped.Params = sp
			}
		}
		return extractComponent(stripped, reqMsg, nil)
	}

	if c.isDerived() {
		return extractDerived(c, msg)
	}
	return extractHeader(c, msg)
}

func extractDerived(c ComponentIdentifier, msg Message) (string, error) {
	switch c.Name {
	case "@method":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @method on response", ErrMissingComponent)
		}
		return strings.ToUpper(msg.Method()), nil

	case "@target-uri":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @target-uri on response", ErrMissingComponent)
		}
		return msg.URL().String(), nil

	case "@authority":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @authority on response", ErrMissingComponent)
		}
		return extractAuthority(msg.URL()), nil

	case "@scheme":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @scheme on response", ErrMissingComponent)
		}
		return strings.ToLower(msg.URL().Scheme), nil

	case "@request-target":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @request-target on response", ErrMissingComponent)
		}
		return extractRequestTarget(msg.URL()), nil

	case "@path":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @path on response", ErrMissingComponent)
		}
		path := msg.URL().Path
		if path == "" {
			path = "/"
		}
		return path, nil

	case "@query":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @query on response", ErrMissingComponent)
		}
		return extractQuery(msg.URL()), nil

	case "@query-param":
		if !msg.IsRequest() {
			return "", fmt.Errorf("%w: @query-param on response", ErrMissingComponent)
		}
		name := c.paramString("name")
		if name == "" {
			return "", fmt.Errorf("%w: @query-param requires ;name parameter", ErrMissingComponent)
		}
		return extractQueryParam(msg.URL(), name)

	case "@status":
		if msg.IsRequest() {
			return "", fmt.Errorf("%w: @status on request", ErrMissingComponent)
		}
		return fmt.Sprintf("%d", msg.StatusCode()), nil

	default:
		return "", fmt.Errorf("%w: unknown derived component %s", ErrMissingComponent, c.Name)
	}
}

func extractHeader(c ComponentIdentifier, msg Message) (string, error) {
	if c.hasParam("sf") {
		return extractHeaderSF(c, msg)
	}
	if c.hasParam("bs") {
		return extractHeaderBS(c, msg)
	}
	if c.hasParam("key") {
		return extractHeaderKey(c, msg)
	}

	values := msg.HeaderValues(c.Name)
	if len(values) == 0 {
		return "", fmt.Errorf("%w: header %q not present", ErrMissingComponent, c.Name)
	}

	// Combine multiple header values with ", " per RFC 9421 Section 2.1
	var trimmed []string
	for _, v := range values {
		trimmed = append(trimmed, strings.TrimSpace(v))
	}
	return strings.Join(trimmed, ", "), nil
}

// extractHeaderSF handles ;sf (structured field) parameter.
// For our purposes, we parse and re-serialize the header value.
func extractHeaderSF(c ComponentIdentifier, msg Message) (string, error) {
	values := msg.HeaderValues(c.Name)
	if len(values) == 0 {
		return "", fmt.Errorf("%w: header %q not present", ErrMissingComponent, c.Name)
	}
	// Re-join and return as-is. Full SFV re-serialization is complex.
	// For RFC 9421 compliance, the value should be parsed as SFV and re-serialized.
	// For now, we trim and join which handles the common case.
	var trimmed []string
	for _, v := range values {
		trimmed = append(trimmed, strings.TrimSpace(v))
	}
	return strings.Join(trimmed, ", "), nil
}

// extractHeaderBS handles ;bs (byte sequence) parameter.
func extractHeaderBS(c ComponentIdentifier, msg Message) (string, error) {
	values := msg.HeaderValues(c.Name)
	if len(values) == 0 {
		return "", fmt.Errorf("%w: header %q not present", ErrMissingComponent, c.Name)
	}
	// ;bs wraps each header field value in SFV Binary Sequence encoding
	var encoded []string
	for _, v := range values {
		encoded = append(encoded, ":"+base64.StdEncoding.EncodeToString([]byte(strings.TrimSpace(v)))+":")
	}
	return strings.Join(encoded, ", "), nil
}

// extractHeaderKey handles ;key for Dictionary Structured Fields.
func extractHeaderKey(c ComponentIdentifier, msg Message) (string, error) {
	key := c.paramString("key")
	if key == "" {
		return "", fmt.Errorf("%w: ;key parameter is empty", ErrMissingComponent)
	}
	values := msg.HeaderValues(c.Name)
	if len(values) == 0 {
		return "", fmt.Errorf("%w: header %q not present", ErrMissingComponent, c.Name)
	}
	combined := strings.Join(values, ", ")
	members, err := ParseDictionary(combined)
	if err != nil {
		return "", fmt.Errorf("%w: cannot parse %q as dictionary: %v", ErrMissingComponent, c.Name, err)
	}
	for _, m := range members {
		if m.Key == key {
			if m.InnerList != nil {
				return serializeInnerList(m.InnerList), nil
			}
			if m.Item != nil {
				result := serializeBareItem(m.Item.Value)
				if m.Item.Params != nil && m.Item.Params.Len() > 0 {
					result += serializeParams(m.Item.Params)
				}
				return result, nil
			}
		}
	}
	return "", fmt.Errorf("%w: key %q not found in dictionary header %q", ErrMissingComponent, key, c.Name)
}

func extractAuthority(u *url.URL) string {
	host := u.Hostname()
	port := u.Port()
	if port == "" {
		return strings.ToLower(host)
	}
	// Omit default ports
	if (u.Scheme == "http" && port == "80") || (u.Scheme == "https" && port == "443") {
		return strings.ToLower(host)
	}
	return strings.ToLower(host) + ":" + port
}

func extractRequestTarget(u *url.URL) string {
	path := u.RequestURI()
	if path == "" {
		return "/"
	}
	return path
}

func extractQuery(u *url.URL) string {
	if u.RawQuery == "" && u.ForceQuery {
		return "?"
	}
	if u.RawQuery == "" {
		return "?"
	}
	return "?" + u.RawQuery
}

func extractQueryParam(u *url.URL, name string) (string, error) {
	values := u.Query()[name]
	if len(values) == 0 {
		return "", fmt.Errorf("%w: query parameter %q not present", ErrMissingComponent, name)
	}
	// RFC 9421: use the first value
	return values[0], nil
}
